Enable basic watching of geolocation

dritter opened the pull request:
Hi there, this is a draft for implementing `geolocation.watchPosition()`.

This is not yet complete, as this has several flaws:
- [x] The `watch` prop is not fully reactive, as `getPosition` has to be toggled to reenable watching position.
- [x] Not documented yet
- [x] clear watcher on destroy

And generally speaking: having two booleans that trigger the geolocation API feels odd. But I am not sure what the best way is to solve this. Could be by giving `getPosition` more meaning (making it a String).
Another idea (not sure if that would work) is to provide a slot for the getPosition-Button, and another one for the watch-Button and listen to the click event (like [here](https://github.com/sveltejs/sapper/issues/731#issuecomment-500487405)).. But I think this would be a dirty solution..

WDYT?

dritter committed:
Enable basic watching of geolocation

metonym commented:
Supporting the `watchPosition` API would be a great feature.

> And generally speaking: having two booleans that trigger the geolocation API feels odd.

What's your use case? Do you need to both get/watch APIs, or just one?

I think that exporting functions to trigger each of the APIs might be worth exploring. This would enable the consumer to programmatically trigger the API without the booleans.

```svelte
<script>
  let ref;

  $: if (ref) ref.getPosition();
</script>

<Geolocation bind:this={ref} />
```

dritter commented:
> What's your use case? Do you need to both get/watch APIs, or just one?

I just need watch, but in my current implementation I would need to set `getPosition` to trigger it initially.

> This would enable the consumer to programmatically trigger the API without the booleans.

That is a great idea. I'll try that in the next days.

dritter committed:
Improve codebase by getting rid of booleans

Additionally, split watchPosition and getPosition into different functions for clarity.

dritter committed:
Automatically cleanup Watcher if Components gets destroyed

dritter commented:
Okay. I just reworked it towards exporting the functions. That feels better to me, with the downside of having to do the wiring from outside is still a bit odd.

Another idea came into my mind: Why not split it into different components? That would make it clear to the dev how to use them (could be `<GeolocationTrackPosition>` and `<GeolocationGetPosition>` or similar) without having to do the wiring.

metonym commented:
> Why not split it into different components? That would make it clear to the dev how to use them (could be <GeolocationTrackPosition> and <GeolocationGetPosition> or similar) without having to do the wiring

Do you mean having separate components while using a boolean as a trigger? Or try this new programmatic approach?

dritter commented:
Neither ;)

If I understand correctly, the boolean trigger is there only to avoid the "Only request geolocation information in response to a user gesture" Violation Warning. This is a fancy way of doing:
```svelte
<script>
  let getPosition = false;
</script>
<button on:click={getPosition = true}>Find Stores near me</button>
{#if getPosition}
  <Geolocation>
{/if}
```
So if this is properly documented, we could just remove the booleans, or am I missing something?

Maybe the component is then simple enough that it could have a `watch` prop (so that we keep a single component). If not, we could split them.

Sorry for bringing up new ideas.

metonym commented:
I'd prefer keeping both methods:

- boolean as a trigger
- exporting a function

dritter committed:
Support getPosition as boolean main switch

dritter committed:
Add documentation

dritter commented:
Okay. Let's keep it then.

PTAL

metonym commented on src/Geolocation.svelte:
```diff
@@ -43,10 +46,12 @@
   export let notSupported = false;
 
-  import { createEventDispatcher } from "svelte";
+  import { createEventDispatcher, onDestroy } from 'svelte'
```
```suggestion
  import { createEventDispatcher, onDestroy } from "svelte";
```

Double quotes and a semicolon

metonym commented on src/Geolocation.svelte:
```diff
@@ -43,10 +46,12 @@
   const dispatch = createEventDispatcher();
 
+  let watcherId = undefined;
+  let lastPosition = undefined;
```
```suggestion
  let lastPosition = undefined;
  
```

Blank line

metonym reviewed (changes requested):
The TypeScript definitions need to be updated as well

dritter committed:
Fix code style

dritter committed:
Add missing TypeScript definitions

dritter commented:
Okay, I added your suggestions. Not sure if I did everything right with the typescript definitions.

metonym commented:
@dritter I should have been more explicit. The type definitions are located in this file: https://github.com/metonym/svelte-geolocation/blob/master/types/Geolocation.d.ts

dritter committed:
Add proper TypeScript definitions

dritter commented:
Ha. Thanks for your guidance. I added the `watch` definition and updated the definition for `getPosition`.

metonym committed:
fix documentation in README

metonym committed:
test new API

metonym committed:
chore: run prettier

metonym reviewed (approved):
Great work. Thanks for taking this on.

metonym commented:
Published v0.2.0 to NPM.
